src/jsCodegen/emitFunctionCallExpr.ts

```diff
@@ -13,121 +30,286 @@ const supportedFunctions: Record<
+	const arity = argAsts.length;
+	const functionNameAndArity = `${localName}/${arity}`;
+
+	// First check if we have a codegen-specific implementation
+	const isDefaultFunctionNamespace = namespaceURI === context.defaultFunctionNamespaceUri;
+	if (isDefaultFunctionNamespace) {
+		const emitFunctionCall = supportedFunctions[functionNameAndArity];
+		if (emitFunctionCall !== undefined) {
+			return emitFunctionCall(ast, contextItemExpr, context);
+		}
 	}
 
-	return rejectAst(`Unsupported function: ${functionName}`);
-}
+	// Check allowlist
+	const allowlist = supportedBuiltinFunctions[isDefaultFunctionNamespace ? '' : namespaceURI];
+	if (allowlist) {
+		// Function is built-in and must be allowed explicitly to be considered safe for codegen
+		if (!allowlist.includes(functionNameAndArity)) {
+			return rejectAst(
+				`Not supported: built-in function not on allow list: ${functionNameAndArity}`
+			);
+		}
+	}
+
+	// Then check registered functions
+	const functionProperties = getFunctionByArity(namespaceURI, localName, arity);
+	if (!functionProperties) {
+		return rejectAst(`Unknown function/arity: ${functionNameAndArity}`);
+	}
 
-const contextItemCheck = `if (contextItem === undefined || contextItem === null) {
-	throw new Error('XPDY0002: The function which was called depends on dynamic context, which is absent.');
-}`;
+	// Should never happen, but let's check anyway
+	if (functionProperties.isUpdating) {
+		return rejectAst('Not supported: updating functions');
+	}
+
+	const argsExpr = emitFunctionArgumentsConversion(
+		argAsts,
+		functionProperties.argumentTypes,
+		contextItemExpr,
+		context
+	);
 
-function createLocalNameGetter(itemName: string) {
-	return `(${itemName}.localName || ${itemName}.target || '')`;
+	const callFunction = mapPartialCompilationResult(argsExpr, (argsExpr) =>
+		acceptAst(
+			`runtimeLib.callFunction(domFacade, ${escapeJavaScriptString(
+				namespaceURI
+			)}, ${escapeJavaScriptString(localName)}, [${argsExpr.code}], options)`,
+			{ type: GeneratedCodeBaseType.Value },
+			argsExpr.variables
+		)
+	);
+	return emitFunctionReturnTypeConversion(callFunction, functionProperties.returnType, context);
 }
 
-function createNameGetter(itemName: string) {
-	return `(((${itemName}.prefix || '').length !== 0 ? ${itemName}.prefix + ':' : '')
-		+ (${itemName}.localName || ${itemName}.target || ''))`;
+function emitContextItemCheck(
+	contextItemExpr: PartialCompilationResult,
+	context: CodeGenContext
+): PartialCompilationResult {
+	return mapPartialCompilationResult(
+		context.getIdentifierFor(contextItemExpr, 'contextItem'),
+		(contextItemExpr) =>
+			acceptAst(contextItemExpr.code, { type: GeneratedCodeBaseType.Value }, [
+				...contextItemExpr.variables,
+				`if (${contextItemExpr.code} === undefined || ${contextItemExpr.code} === null) {
+					throw new Error('XPDY0002: The function which was called depends on dynamic context, which is absent.');
```

Comment:
Why are you not using the runtimelib here?

Reply:
To this and the above - error throwing was and still is a bit inconsistent - do you prefer to import all error helpers through runtimeLib or to turn everything into `new Error('XP...')`?